backend/src/apiserver/resource/resource_manager.go

```diff
@@ -1058,3 +1106,22 @@ func (r *ResourceManager) GetNamespaceFromRunID(runId string) (string, error) {
 	}
 	return namespace, nil
 }
+
+func (r *ResourceManager) GetNamespaceFromJobID(jobId string) (string, error) {
+	job, err := r.GetJob(jobId)
+	if err != nil {
+		return "", util.Wrap(err, "Failed to get namespace from Job ID.")
+	}
+	namespace := job.Namespace
```

Comment:
I think you can trust job.Namespace is not empty and report an error directly, because it is synced in https://github.com/kubeflow/pipelines/blob/3d8077cc37779efc42aa09439f7ac74e12cb5b4b/backend/src/apiserver/storage/job_store.go#L346. persistence agent has a loop to send swf CRs into pipeline reportScheduledWorkflow api method. The CRs will always have namespace, so data in db will always have them similar to workflows.

Reply:
Thanks! A quick question: Is persistence agent always deployed? also does this apply to run as well? I got this code from run, per existing comment, it says legacy runs may have empty namespace. Wonder if that still apply in certain cases.

Reply:
discussed offline, from commit history, looks like Namespace was stored at the beginning when KFP was release publically in Github. So it would not be a risky choice.

Reply:
Done.